### PR TITLE
Enhancing returned errors for specialresource controller.

### PR DESCRIPTION
**Pull request by ybettan**
This is part of the logging enhancement effort which says that:
  * all errors should be wrapped with its context
  * error logging should only be done at the reconciliation loop level
  * info and warning logs can be everywhere in the code

Signed-off-by: Yoni Bettan <yonibettan@gmail.com>

### CODE REVIEW

File: controllers/specialresource.go [hunk 1]
```diff
@@ -117,26 +122,26 @@ func (r *SpecialResourceReconciler) SpecialResourcesReconcile(ctx context.Contex
...
+		log.Info(utils.WarnString("RECONCILE REQUEUE: Could not reconcile chart"), "error", err)
 		return reconcile.Result{Requeue: true}, nil
 	}
 
 	if suErr := r.StatusUpdater.SetAsReady(ctx, wi.SpecialResource, state.Success, ""); suErr != nil {
-		log.Error(suErr, "Failed to update CR's status to Ready")
-		return reconcile.Result{}, suErr
+		return reconcile.Result{}, fmt.Errorf("failed to update CR's status to Ready: %w", suErr)
 	}
 	log.Info("RECONCILE SUCCESS")
```
Comment (yevgeny-shnaidman) [inline controllers/specialresource.go:146]: This log can be deleted, we have a log in the Reconcile function that signifies the end of the loop

File: controllers/specialresource.go [hunk 2]
```diff
@@ -56,19 +62,18 @@ func (r *SpecialResourceReconciler) SpecialResourcesReconcile(ctx context.Contex
...
+		return reconcile.Result{}, fmt.Errorf("Failed to update SpecialResource's %s/%s status to Progressing: %w", wi.SpecialResource.Namespace, wi.SpecialResource.Name, err)
 	}
 
 	var err error
 	wi.Chart, err = r.Helmer.Load(wi.SpecialResource.Spec.Chart)
 	if err != nil {
-		log.Error(err, "Failed to load the Helm chart")
-		if suErr := r.StatusUpdater.SetAsErrored(ctx, wi.SpecialResource, state.ChartFailure, fmt.Sprintf("Failed to load Helm Chart: %v", err)); suErr != nil {
-			log.Error(suErr, "Failed to update CR's status to Errored")
+		msg := "failed to load helm chart"
```
Comment (yevgeny-shnaidman) [inline controllers/specialresource.go:72]: lets add the name of the chart to the message

File: controllers/specialresource.go [hunk 3]
```diff
@@ -80,12 +85,12 @@ func (r *SpecialResourceReconciler) SpecialResourcesReconcile(ctx context.Contex
 		cctx := ctrl.LoggerInto(ctx, clog)
 
 		cchart, err := r.Helmer.Load(dependency.HelmChart)
-		log.Error(err, "Failed to load dependency Helm chart")
 		if err != nil {
-			if suErr := r.StatusUpdater.SetAsErrored(cctx, wi.SpecialResource, state.DependencyChartFailure, fmt.Sprintf("Failed to load dependency Helm Chart: %v", err)); suErr != nil {
-				clog.Error(suErr, "Failed to update CR's status to Errored")
+			msg := "failed to load dependency Helm Chart"
```
Comment (yevgeny-shnaidman) [inline controllers/specialresource.go:89]: lets add the name of the chart to the message

File: controllers/specialresource.go [hunk 4]
```diff
@@ -117,26 +122,26 @@ func (r *SpecialResourceReconciler) SpecialResourcesReconcile(ctx context.Contex
...
 	log.Info("Done resolving dependencies - reconciling main SpecialResource")
 	if err := r.ReconcileSpecialResourceChart(ctx, wi); err != nil {
-		if suErr := r.StatusUpdater.SetAsErrored(ctx, wi.SpecialResource, state.FailedToDeployChart, fmt.Sprintf("Failed to deploy SpecialResource's chart: %v", err)); suErr != nil {
-			log.Error(suErr, "failed to update CR's status to Errored")
+		msg := "failed to deploy SpecialResource's chart"
+		if suErr := r.StatusUpdater.SetAsErrored(ctx, wi.SpecialResource, state.FailedToDeployChart, fmt.Sprintf("%s: %v", msg, err)); suErr != nil {
+			log.Info(utils.WarnString("failed to update CR's status to Errored"), "error", suErr)
 		}
-		log.Error(err, "RECONCILE REQUEUE: Could not reconcile chart")
+		log.Info(utils.WarnString("RECONCILE REQUEUE: Could not reconcile chart"), "error", err)
```
Comment (yevgeny-shnaidman) [inline controllers/specialresource.go:139]: lets add a name of the chart

**Comment by yevgeny-shnaidman** [comment]
what about specialresource_controller.go file? There are a couple of logs that need to be removed/changed, and maybe added

File: controllers/specialresource.go (hunk 4 above)
Comment (ybettan) [reply controllers/specialresource.go]: Fixed.

File: controllers/specialresource.go (hunk 2 above)
Comment (ybettan) [reply controllers/specialresource.go]: Fixed.

File: controllers/specialresource.go (hunk 1 above)
Comment (ybettan) [reply controllers/specialresource.go]: Fixed.

**Comment by ybettan** [comment]
@yevgeny-shnaidman PTAL

**Comment by ybettan** [comment]
@yevgeny-shnaidman PTAL

**Comment by ybettan** [comment]
@yevgeny-shnaidman PTAL

File: controllers/specialresource_controller.go [hunk 5]
```diff
@@ -173,5 +172,8 @@ func (r *SpecialResourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			RateLimiter:             workqueue.NewItemExponentialFailureRateLimiter(minDelaySR, maxDelaySR),
 		}).
 		WithEventFilter(r.Filter.GetPredicates()).
-		Complete(r)
+		Complete(r); err != nil {
+		return errors.New("couldn't create 'specialresource' controller")
```
[the remark this answers is not in the document]
Comment (ybettan) [reply controllers/specialresource_controller.go]: It looks wired cause I forgot to wrap the error. It is fixed now.

[the remark this answers is not in the document]
Comment (ybettan) [reply controllers/specialresource_controller.go]: I agree that there is no much benefit but we agreed that all errors should be wrapped so I don't think we should make an exception here.

File: controllers/specialresource.go [hunk 6]
```diff
@@ -117,66 +122,68 @@ func (r *SpecialResourceReconciler) SpecialResourcesReconcile(ctx context.Contex
...
 	log.Info("Done resolving dependencies - reconciling main SpecialResource")
 	if err := r.ReconcileSpecialResourceChart(ctx, wi); err != nil {
-		if suErr := r.StatusUpdater.SetAsErrored(ctx, wi.SpecialResource, state.FailedToDeployChart, fmt.Sprintf("Failed to deploy SpecialResource's chart: %v", err)); suErr != nil {
-			log.Error(suErr, "failed to update CR's status to Errored")
+		msg := "failed to deploy SpecialResource's chart"
+		if suErr := r.StatusUpdater.SetAsErrored(ctx, wi.SpecialResource, state.FailedToDeployChart, fmt.Sprintf("%s: %v", msg, err)); suErr != nil {
+			log.Info(utils.WarnString("failed to update CR's status to Errored"), "error", suErr)
 		}
-		log.Error(err, "RECONCILE REQUEUE: Could not reconcile chart")
+		log.Info(utils.WarnString("RECONCILE REQUEUE: Could not reconcile chart for SpecialResource"), "specialresource", wi.SpecialResource.Name, "error", err)
```
Comment (yevgeny-shnaidman) [inline controllers/specialresource.go:139]: why do we need to supply special resource? it is part of the log configuration

File: controllers/specialresource_controller.go [hunk 7]
```diff
@@ -110,10 +109,9 @@ func (r *SpecialResourceReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 
 	// Reconcile all specialresources
 	if res, err = r.SpecialResourcesReconcile(ctx, wi); err != nil || res.Requeue {
-		return res, errors.Wrap(err, "Failed to reconcile SpecialResource")
+		return res, fmt.Errorf("failed to reconcile SpecialResource '%s/%s': %w", wi.SpecialResource.Namespace, wi.SpecialResource.Name, err)
 	}
 
-	log.Info("Reconciliation successful")
```
Comment (yevgeny-shnaidman) [inline controllers/specialresource_controller.go:116]: we want to keep this message, it signals the successful end of reconcilaition.

File: controllers/specialresource.go (hunk 6 above)
Comment (ybettan) [reply controllers/specialresource.go]: Fixed.

File: controllers/specialresource_controller.go (hunk 7 above)
Comment (ybettan) [reply controllers/specialresource_controller.go:116]: Retrieved.

**Comment by ybettan** [comment]
@yevgeny-shnaidman PTAL

**Pull request merged**